core/src/main/java/com/yahoo/imapnio/async/request/AbstractStoreFlagsCommand.java

```diff
@@ -108,23 +146,58 @@ protected AbstractStoreFlagsCommand(final boolean isUid, @Nonnull final String m
         this.flags = flags;
         this.action = action;
         this.isSilent = silent;
+        this.unchangedSince = null;
+    }
+
+    /**
+     * Initializes a {@link AbstractStoreFlagsCommand} with string form message numbers (could be sequence sets or UIDs), flags, action,
+     * silent flag whether server should return new values, and unchanged since the given modification sequence.
+     *
+     * @param isUid whether to have UID prepended
+     * @param msgNumbers the message id
+     * @param flags the flags to be stored
+     * @param action whether to replace, add or remove the flags
+     * @param silent true if asking server to respond silently
+     * @param unchangedSince unchanged since the given modification sequence
+     */
+    protected AbstractStoreFlagsCommand(final boolean isUid, @Nonnull final String msgNumbers, @Nonnull final Flags flags,
+            @Nonnull final FlagsAction action, final boolean silent, @Nonnull final Long unchangedSince) {
+        this.isUid = isUid;
+        this.msgNumbers = msgNumbers;
+        this.flags = flags;
+        this.action = action;
+        this.isSilent = silent;
+        this.unchangedSince = unchangedSince;
     }
 
     @Override
     public void cleanup() {
         this.msgNumbers = null;
         this.flags = null;
         this.action = null;
+        this.unchangedSince = null;
     }
 
     @Override
     public ByteBuf getCommandLineBytes() {
         // Ex:STORE 2:4 +FLAGS (\Deleted)
-        final ByteBuf sb = Unpooled.buffer();
+        final ImapArgumentFormatter argWriter = new ImapArgumentFormatter();
+        final String flagListString = argWriter.buildFlagString(flags);
+        final String unchangedSinceStr = unchangedSince == null ? "" : unchangedSince.toString();
+        final int len = ImapClientConstants.PAD_LEN  + msgNumbers.length() + unchangedSinceStr.length() + flagListString.length();
+        final ByteBuf sb = Unpooled.buffer(len);
         sb.writeBytes(isUid ? UID_STORE_SP_B : STORE_SP_B);
         sb.writeBytes(msgNumbers.getBytes(StandardCharsets.US_ASCII));
         sb.writeByte(ImapClientConstants.SPACE);
 
+        if (unchangedSince != null) {
+            sb.writeByte(ImapClientConstants.L_PAREN);
+            sb.writeBytes(UNCHANGEDSINCE_SP_B);
+            sb.writeBytes(unchangedSinceStr.getBytes(StandardCharsets.US_ASCII));
```

Comment:
It seems that the server side only accept ByteBuff from String.

Reply:
we cannot use writeLong method in ByteBuf is because  ByteBuf will write it in binary format, whereas IMAP protocol only understands  ascii code 0 - 9.  Aka, following is the ABNF

https://tools.ietf.org/html/rfc5234
A range of alternative numeric values can be specified compactly,
   using a dash ("-") to indicate the range of alternative values.
   Hence:
         DIGIT       =  %x30-39

   is equivalent to:
         DIGIT       =  "0" / "1" / "2" / "3" / "4" / "5" / "6" /
                        "7" / "8" / "9"